multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +906,331 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
 
 	return nil
 }
+
+func (rcsw *RemoteClusterServiceWatcher) createOrUpdateHeadlessEndpoints(ctx context.Context, remoteEp *corev1.Endpoints) error {
+	// Get local version of the Endpoints object and diff
+	localEpName := rcsw.mirroredResourceName(remoteEp.Name)
+	localEndpoints, err := rcsw.localAPIClient.Endpoint().Lister().Endpoints(remoteEp.Namespace).Get(localEpName)
+	// If Endpoints update on the remote cluster, and the mirrorerd version of
+	// the Endpoints does not exist then it means we need to create it (and
+	// all of its nested clusterIP services).
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		}
+
+		if err := rcsw.handleHeadlessEndpointsCreated(ctx, remoteEp); err != nil {
+			rcsw.log.Debugf("failed to create headless mirrors for Endpoints %s/%s: %v", remoteEp.Namespace, remoteEp.Name, err)
+			return err
+		}
+
+		return nil
+	}
+
+	localEp := localEndpoints.DeepCopy()
+	remoteService, err := rcsw.remoteAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(remoteEp.Name)
+	if err != nil {
+		rcsw.log.Debugf("failed to retrieve remote service %s/%s when updating its headless Endpoints: %v", remoteEp.Namespace, remoteEp.Name, err)
+		return fmt.Errorf("could not retrieve remote service during headless Endpoints update: %s", err)
+	}
+
+	// First, get all hostnames
+	localServiceNames := make(map[string]struct{})
+	newSubsets := make([]corev1.EndpointSubset, 0, len(remoteEp.Subsets))
+	for _, subset := range remoteEp.Subsets {
+		newAddresses := make([]corev1.EndpointAddress, 0, len(subset.Addresses))
+		for _, address := range subset.Addresses {
+			if address.TargetRef == nil {
```

Comment:
As in my other comment, I think we should be working off the address.Hostname instead of the targetRef